src/config/config.c

```diff
@@ -0,0 +1,256 @@
+#include "gui/gui.h"
+#include "config.h"
+#include "display.h"
+#include "mixer.h"
+#include "tx.h"
+
+#include <stdlib.h>
+#include <string.h>
+
+#define MATCH_KEY(s)     strcasecmp(name,    s) == 0
+#define MATCH_VALUE(s)   strcasecmp(value,   s) == 0
+
+typedef const char* StringCallback(int value);
+typedef const char* StringCallback2(char* buffer, int value);
+
+extern u8 FONT_GetFromString(const char *);
+
+static u16 get_color(const char *value) {
+    u8 r, g, b;
+    u32 color = strtol(value, NULL, 16);
+    r = 0xff & (color >> 16);
+    g = 0xff & (color >> 8);
+    b = 0xff & (color >> 0);
+    return RGB888_to_RGB565(r, g, b);
+}
+
+static u8 get_button(const char *value)
+{
+    u8 i;
+    for (i = 0; i <= NUM_TX_BUTTONS; i++) {
+        if (strcasecmp(INPUT_ButtonName(i), value) == 0) {
+            return i;
+        }
+    }
+    printf("Could not parse Button %s\n", value);
+    return 0;
+}
+
+
+static s8 mapstrcasecmp(const char *s1, const char *s2)
+{
+    int i = 0;
+    while (1) {
+        if (s1[i] == s2[i]
+           || (s2[i] >= 'a' && s1[i] + ('a'-'A') == s2[i])
+           || (s1[i] >= 'a' && s2[i] + ('a'-'A') == s1[i])
+           || (s2[i] == ' ' && s1[i] == '_')
+           || (s1[i] == ' ' && s2[i] == '_'))
+        {
+            if (s1[i] == '\0')
+                return 0;
+            i++;
+            continue;
+        }
+        return (s1[i] < s2[i] ? -1 : 1);
+    }
+}
+
+u8 get_source(const char *value)
+{
+    unsigned i;
+    unsigned val;
+    const char *ptr = (value[0] == '!') ? value + 1 : value;
+    const char *tmp;
+    char cmp[10];
+    for (i = 0; i <= NUM_SOURCES; i++) {
+        if (mapstrcasecmp(INPUT_SourceNameReal(cmp, i), ptr) == 0) {
+            #if defined(HAS_SWITCHES_NOSTOCK) && HAS_SWITCHES_NOSTOCK
+            #define SWITCH_NOSTOCK ((1 << INP_HOLD0) | (1 << INP_HOLD1) | \
+                                    (1 << INP_FMOD0) | (1 << INP_FMOD1))
+            if ((Transmitter.ignore_src & SWITCH_NOSTOCK) == SWITCH_NOSTOCK) {
+                if (mapstrcasecmp("FMODE0", ptr) == 0 ||
+                    mapstrcasecmp("FMODE1", ptr) == 0 ||
+                    mapstrcasecmp("HOLD0", ptr) == 0 ||
+                    mapstrcasecmp("HOLD1", ptr) == 0)
+                    break;
+            }
+            #endif  // HAS_SWITCHES_NOSTOCK
+            return ((ptr == value) ? 0 : 0x80) | i;
+        }
+    }
+    for (i = 0; i < 4; i++) {
+        if (mapstrcasecmp(tx_stick_names[i], ptr) == 0) {
+            return ((ptr == value) ? 0 : 0x80) | (i + 1);
+        }
+    }
+    i = 0;
+    while ((tmp = INPUT_MapSourceName(i++, &val))) {
+        if (mapstrcasecmp(tmp, ptr) == 0) {
+            return ((ptr == value) ? 0 : 0x80) | val;
+        }
+    }
+    printf("Could not parse Source %s\n", value);
+    return 0;
+}
+
+int assign_int(void* ptr, const struct struct_map *map, int map_size, const char* name, const char* value) {
+    for (int i = 0; i < map_size; i++) {
```

Comment:
is it worth passing 2 extra pointers for every call to assign_int for some debug that is normally disabled?  Handing TYPE_STR_CALL separately and using a wrapper for debug may allow you to save quite a bit of space for each call

Reply:
Not sure if I follow you. name, value are used by macro:
#define MATCH_KEY(s)     strcasecmp(name,    s) == 0
#define MATCH_VALUE(s)   strcasecmp(value,   s) == 0

maybe I should remove this macro as I only use them once.

Reply:
yeah, let's remove the macro as it wasn't obvious to me.  If they were commonly used, that'd be something different.